### ISSUE TITLE
Garbage polygons mixed in with the good ones

### ISSUE THREAD

**Issue by omgitsraven** [libfive/libfive#150]
1) Open the following model: [demo1_debug.zip](https://github.com/libfive/libfive/files/1779051/demo1_debug.zip)
2) Render it out at resolution 100 and quality 11 (it should be by default)
3) Export the .stl
4) Open the .stl in something like Blender, or https://www.viewstl.com/ (with 'edges' set to 'yes')
5) Look carefully at the sharpest edge: there's all kinds of stray polygons doubled up on the surface, some of them apparently facing the wrong way?

![badgeom_mesh](https://user-images.githubusercontent.com/8540042/36951798-a10beafc-1fd5-11e8-8aea-b51e0fa1364b.PNG) ![badgeom_zfight](https://user-images.githubusercontent.com/8540042/36951800-a2f9abc4-1fd5-11e8-9d39-82af65f9eded.PNG)

This seems to be messing up attempts to cleanly decimate the mesh, but honestly I'm not even completely sure; the topology is very confusing...

**Comment by mkeeter** [comment]
Yup, that looks like self-intersection, which is a known issues with the whole Dual Contouring family of algorithms.  At this point, I'd expect meshing to be

- Manifold
- Watertight
- Hierarchical
- Sharp-feature preserving

Self-intersections are definitely still an issue (and came up when I was looking into [smooth rendering](https://github.com/libfive/libfive/issues/45#issuecomment-354207254)).

[This paper](http://www.cs.wustl.edu/~taoju/research/interfree_paper_final.pdf) has a proposed fix, and it's on my list to look into.
